src/codegen/debug.rs

```diff
@@ -351,9 +416,208 @@ impl<'ink> DebugBuilder<'ink> {
+            return_type,
+            &parameter_types,
+            DIFlagsConstants::PUBLIC,
+        ))
+    }
+
+    fn create_function(
+        &self,
+        pou: &PouIndexEntry,
+        return_type: Option<&DataType>,
+        parameter_types: &[&DataType],
+        implementation_start: u32,
+    ) -> Result<DISubprogram, Diagnostic> {
+        let is_external = matches!(pou.get_linkage(), LinkageType::External);
+        let ditype = self.create_subroutine_type(return_type, parameter_types)?;
+        Ok(self.debug_info.create_function(
+            self.compile_unit.get_file().as_debug_info_scope(),
+            pou.get_name(),
+            Some(pou.get_name()), // for generics e.g. NAME__TYPE
+            self.compile_unit.get_file(),
+            (pou.get_location().line_number + 1) as u32, // This sets the start of the POU as the
+            // entry for the function
+            ditype,
+            false, // TODO: what is this
+            !is_external,
+            implementation_start + 1,
+            DIFlagsConstants::PUBLIC,
+            self.optimization.is_optimized(),
+        ))
+    }
+
+    ///Creates the debug information for function variables
+    ///For a `Function` these will be all VAR_INPUT, VAR_OUTPUT and VAR_IN_OUT in addition to
+    ///entries for VAR and VAR_TEMP
+    ///For other POUs we create enties in VAR_TEMP and an additional single parameter at position 0
+    ///(the struct)
+    fn create_function_variables(
+        &mut self,
+        pou: &PouIndexEntry,
+        func: FunctionValue<'ink>,
+        index: &Index,
+    ) -> Result<(), Diagnostic> {
+        let mut param_offset = 0;
+        //Register the return and local variables for debugging
+        for variable in index
+            .get_container_members(pou.get_name())
+            .iter()
+            .filter(|it| it.is_local() || it.is_temp() || it.is_return())
+        {
+            let var_type = index
+                .find_effective_type_by_name(variable.get_type_name())
+                .expect("Type should exist at this stage");
+            let alignment = var_type.get_type_information().get_alignment(index).bits();
+            //If the variable is an aggregate return type, register it as first parameter, and
+            //increase the param count
+            if variable.is_return() && var_type.is_aggregate_type() {
+                self.register_aggregate_return(variable, var_type, func)?;
+                param_offset += 1;
+            } else {
+                self.register_local_variable(variable, alignment, func)?;
+            }
+        }
+        let implementation = pou
+            .find_implementation(index)
+            .expect("A POU will have an impl at this stage");
+        if implementation.implementation_type != ImplementationType::Function {
+            if implementation.get_implementation_type() == &ImplementationType::Method {
+                todo!("Method debugging");
```

Comment:
now codegen will panic for methods?

Reply:
not unless it's being compiled with -g, but I could make this a diagnostic instead.